File: python/xmppcontenthandler.py
import logging
from typing import List
from xml.sax import ContentHandler

from xmppelement import XmppElement

logger = logging.getLogger(__name__)


class XmppContentHandler(ContentHandler):
    def __init__(self):
        ContentHandler.__init__(self)
        self.queue = []  # type: List[XmppElement]
        self.current_element = None
        self.element_stack = []

    def startElement(self, name, attrs):
        element = XmppElement()
        element.tag = name
        element.attributes = attrs

        if name == "stream:stream":
            # This element won't close until stream is closed
            self.queue.append(element)
            return

        self.element_stack.append(self.current_element)
        self.current_element = element

    def endElement(self, name):
        if not self.current_element:
            if name == "stream:stream":
                element = XmppElement()
                element.tag = "stream:closed"
                return
            else:
                raise RuntimeError("No current element")

        if name != self.current_element.tag:
            raise RuntimeError("Mismatched tag")

        parent = self.element_stack.pop()
        if parent:
            parent.children.append(self.current_element)
        else:
            self.queue.insert(0, self.current_element)
        self.current_element = parent

    def characters(self, content):
        if self.current_element.text:
            self.current_element.text += content
        else:
            self.current_element.text = content

File: go/xmpp/xmppcontenthandler.go
package xmpp

import (
	"encoding/xml"
	"bytes"
	"io"
	"strings"
	"errors"
	"container/list"
)

type XmppContentHandler struct {
	queue []*XmppElement
	current_element *XmppElement
	element_stack *list.List
	leftovers string

	// Keep previous State to restore it in the event of an incomplete input
	prev_queue []*XmppElement
	prev_current_element *XmppElement
	prev_element_stack *list.List
}

func NewXmppContentHandler() *XmppContentHandler {
	return &XmppContentHandler{}
}

func (handler *XmppContentHandler) Parse(input string) (err error) {
	buffer := new(bytes.Buffer)
	parser := xml.NewDecoder(buffer)
	var current_input = handler.leftovers + input
	buffer.WriteString(current_input)
	for {
		var token xml.Token
		token, err = parser.Token()
		if err == io.EOF {
			err = nil
			break
		}
		if err != nil && strings.HasSuffix(err.Error(), "unexpected EOF") {
			// Restore State to last know good
			handler.current_element = handler.prev_current_element
			handler.queue = handler.prev_queue
			handler.element_stack = handler.prev_element_stack
			handler.leftovers = current_input

			err = nil
			break
		}
		if err != nil {
			break
		}

		if handler.element_stack == nil {
			handler.element_stack = list.New()
		}

		switch token.(type)  {
		case xml.StartElement:
			start := token.(xml.StartElement)
			element := XmppElement{
				Tag:        start.Name.Local,
				Namespace:  start.Name.Space,
				Attributes: map[string]string{},
			}
			for _, attribute := range start.Attr  {
				element.Attributes[attribute.Name.Local] = attribute.Value
			}
			if element.Tag == "stream" {
	            // This element won't close until stream is closed
				handler.queue = append(handler.queue, &element)
				continue
			}
			handler.element_stack.PushFront(handler.current_element)
			handler.current_element = &element

		case xml.EndElement:
			if handler.current_element == nil {
				err = errors.New("No current element")
				return
			}
			parentListNode := handler.element_stack.Front()
			handler.element_stack.Remove(parentListNode)
			parent := parentListNode.Value.(*XmppElement)
			current_element := handler.current_element
			handler.current_element = parent
			if parent != nil {
				parent.Children = append(parent.Children, current_element)
			} else {
				handler.queue = append(handler.queue, current_element)

				// We've just completed a stanza - store the current State
				// so we can handle incomplete xml snippets
				handler.prev_current_element = handler.current_element
				handler.prev_queue = handler.queue
				handler.prev_element_stack = handler.element_stack

				current_input, _ = buffer.ReadString(0)
				buffer.WriteString(current_input)
			}

		case xml.CharData:
			if handler.current_element != nil {
				handler.current_element.Text = string(token.(xml.CharData))
			}
		}
	}

	return err
}
